File: src/main/java/com/euc/grantfitassessment/eval/DriftExperimentRunner.java
package com.euc.grantfitassessment.eval;

import com.euc.core.EucDefinition;
import com.euc.grantfitassessment.AssessmentResult;
import com.euc.grantfitassessment.GrantFitApplication;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Runs the drift-detection experiment described in docs/proposal.md
 * Section 7: hold the EUC fixed, run the eval dataset against a baseline
 * FitReasoner and one or more candidate variants (model swaps, prompt
 * edits), and measure whether evaluation against the unchanged EUC still
 * tracks the resulting behavior.
 *
 * How "flagged" is operationalized here (see compare()) is a specific
 * design choice, not something the EUC schema dictates: a case counts as
 * flagged when the evaluator agreed with ground truth under the baseline
 * but stopped agreeing under the variant — i.e. evaluation caught a
 * behavior change on that datapoint. A study design that wants a stricter
 * or looser definition should adjust compare() rather than treat this as
 * fixed by the schema.
 */
public class DriftExperimentRunner {

    private final EucDefinition euc;
    private final List<TestCase> dataset;

    public DriftExperimentRunner(EucDefinition euc, List<TestCase> dataset) {
        this.euc = euc;
        this.dataset = dataset;
    }

    public DriftExperimentReport run(FitReasonerVariant baseline, List<FitReasonerVariant> candidates) {
        Map<String, CaseOutcome> baselineOutcomes = scoreVariant(baseline);

        List<VariantResult> variantResults = new ArrayList<>();
        for (FitReasonerVariant candidate : candidates) {
            Map<String, CaseOutcome> candidateOutcomes = scoreVariant(candidate);
            variantResults.add(compare(candidate, baselineOutcomes, candidateOutcomes));
        }

        return DriftExperimentReport.from(baseline, variantResults);
    }

    private Map<String, CaseOutcome> scoreVariant(FitReasonerVariant variant) {
        GrantFitApplication app = new GrantFitApplication(euc, variant.reasoner());
        GrantFitEvaluator evaluator = new GrantFitEvaluator(euc);

        Map<String, CaseOutcome> outcomes = new LinkedHashMap<>();
        for (TestCase tc : dataset) {
            AssessmentResult actual = app.assess(tc.organization(), tc.grant());
            GrantFitEvaluator.EvaluationScore score = evaluator.evaluate(tc, actual);
            outcomes.put(tc.caseId(), new CaseOutcome(tc.caseId(), actual, score));
        }
        return outcomes;
    }

    private VariantResult compare(FitReasonerVariant candidate,
                                   Map<String, CaseOutcome> baselineOutcomes,
                                   Map<String, CaseOutcome> candidateOutcomes) {
        List<String> flaggedCases = new ArrayList<>();
        List<String> eligibilityCorrectnessDriftCases = new ArrayList<>();
        int eligibleCases = 0;
        int groundedEligibleCases = 0;

        for (String caseId : baselineOutcomes.keySet()) {
            CaseOutcome baselineOutcome = baselineOutcomes.get(caseId);
            CaseOutcome candidateOutcome = candidateOutcomes.get(caseId);

            if (baselineOutcome.score().allPassed() && !candidateOutcome.score().allPassed()) {
                flaggedCases.add(caseId);
            }

            // deterministic-rule-stability: eligibilityCorrectness must not move
            // when only the reasoning layer (alignmentReasoning) changes, since
            // the deterministic filters (eligibility/geography/requiredInfo) are
            // untouched by any FitReasoner variant.
            if (baselineOutcome.score().eligibilityCorrectness() != candidateOutcome.score().eligibilityCorrectness()) {
                eligibilityCorrectnessDriftCases.add(caseId);
            }

            if (candidateOutcome.actual().eligible()) {
                eligibleCases++;
                if (candidateOutcome.score().evidenceGrounding()) {
                    groundedEligibleCases++;
                }
            }
        }

        double evidenceGroundingRate = eligibleCases == 0 ? Double.NaN : (double) groundedEligibleCases / eligibleCases;

        return new VariantResult(candidate, flaggedCases, eligibilityCorrectnessDriftCases,
                evidenceGroundingRate, candidateOutcomes);
    }

    /** One test case's outcome for a given variant: the raw result plus its score against ground truth. */
    public record CaseOutcome(String caseId, AssessmentResult actual, GrantFitEvaluator.EvaluationScore score) {
    }

    /** One candidate variant's comparison against the baseline, across the full dataset. */
    public record VariantResult(
            FitReasonerVariant variant,
            List<String> flaggedCaseIds,
            List<String> eligibilityCorrectnessDriftCaseIds,
            double evidenceGroundingRate,
            Map<String, CaseOutcome> outcomes
    ) {
        public boolean anyDriftFlagged() {
            return !flaggedCaseIds.isEmpty();
        }

        public boolean deterministicRuleStable() {
            return eligibilityCorrectnessDriftCaseIds.isEmpty();
        }
    }
}

File: src/euc/grantfitassessment/eval/drift_experiment_runner.py
from __future__ import annotations

import math
from dataclasses import dataclass, field

from euc.core.models import EucDefinition
from euc.grantfitassessment.app import GrantFitApplication
from euc.grantfitassessment.eval.dataset import TestCase
from euc.grantfitassessment.eval.evaluator import EvaluationScore, GrantFitEvaluator
from euc.grantfitassessment.eval.fit_reasoner_variant import FitReasonerVariant
from euc.grantfitassessment.models import AssessmentResult


@dataclass(frozen=True)
class CaseOutcome:
    case_id: str
    actual: AssessmentResult
    score: EvaluationScore


@dataclass(frozen=True)
class VariantResult:
    variant: FitReasonerVariant
    flagged_case_ids: list[str]
    eligibility_correctness_drift_case_ids: list[str]
    evidence_grounding_rate: float
    outcomes: dict[str, CaseOutcome] = field(default_factory=dict)

    def any_drift_flagged(self) -> bool:
        return len(self.flagged_case_ids) > 0

    def deterministic_rule_stable(self) -> bool:
        return len(self.eligibility_correctness_drift_case_ids) == 0


class DriftExperimentRunner:
    def __init__(self, euc: EucDefinition, dataset: list[TestCase]) -> None:
        self._euc = euc
        self._dataset = dataset

    def run(self, baseline: FitReasonerVariant, candidates: list[FitReasonerVariant]):
        from euc.grantfitassessment.eval.drift_experiment_report import DriftExperimentReport

        baseline_outcomes = self._score_variant(baseline)
        variant_results = []
        for candidate in candidates:
            candidate_outcomes = self._score_variant(candidate)
            variant_results.append(self._compare(candidate, baseline_outcomes, candidate_outcomes))
        return DriftExperimentReport.from_results(baseline, variant_results)

    def _score_variant(self, variant: FitReasonerVariant) -> dict[str, CaseOutcome]:
        app = GrantFitApplication(self._euc, variant.reasoner)
        evaluator = GrantFitEvaluator(self._euc)

        outcomes: dict[str, CaseOutcome] = {}
        for test_case in self._dataset:
            actual = app.assess(test_case.organization, test_case.grant)
            score = evaluator.evaluate(test_case, actual)
            outcomes[test_case.case_id] = CaseOutcome(test_case.case_id, actual, score)
        return outcomes

    def _compare(
        self,
        candidate: FitReasonerVariant,
        baseline_outcomes: dict[str, CaseOutcome],
        candidate_outcomes: dict[str, CaseOutcome],
    ) -> VariantResult:
        flagged_cases: list[str] = []
        eligibility_correctness_drift_cases: list[str] = []
        eligible_cases = 0
        grounded_eligible_cases = 0

        for case_id in baseline_outcomes:
            baseline_outcome = baseline_outcomes[case_id]
            candidate_outcome = candidate_outcomes[case_id]

            if baseline_outcome.score.all_passed() and not candidate_outcome.score.all_passed():
                flagged_cases.append(case_id)

            if baseline_outcome.score.eligibility_correctness != candidate_outcome.score.eligibility_correctness:
                eligibility_correctness_drift_cases.append(case_id)

            if candidate_outcome.actual.eligible:
                eligible_cases += 1
                if candidate_outcome.score.evidence_grounding:
                    grounded_eligible_cases += 1

        evidence_grounding_rate = (
            math.nan if eligible_cases == 0 else grounded_eligible_cases / eligible_cases
        )

        return VariantResult(
            variant=candidate,
            flagged_case_ids=flagged_cases,
            eligibility_correctness_drift_case_ids=eligibility_correctness_drift_cases,
            evidence_grounding_rate=evidence_grounding_rate,
            outcomes=candidate_outcomes,
        )
